### PR TITLE
fix(wandb): survive setup_wandb() being called twice in one job

### PR DESCRIPTION
**Pull request by saforem2**
## Regression from #201 (v0.24.2) — reported from a live Sunspot run

Creating the W&B run early in `main()` made `setup_wandb` run a **second** time when `History`'s `WandbBackend` later adopted it. Nothing had exercised that path before.

```
wandb: wandb.init() called while a run is active and reinit is set to 'default', so returning the previous run.
[...] wandb.run=[devout-totem-1112](https://wandb.ai/aurora_gpt/.../sdkpnx4k)
[...][E][ezpz/distributed:1705:setup_wandb] wandb.init() failed from rank=0:
    Attempted to change value of key "tstamp"
    from 2026-08-05T16:56:09.118208 to 2026-08-05T16:56:57.869198
[...][W][ezpz/distributed:1706:setup_wandb] Continuing without wandb logging.
```

### Why this is worse than the bug it replaced

The post-init `run.config.update({...})` (`distributed.py:1664`) writes wall-clock keys — `tstamp`, `year`/`month`/`day` — and passed **no** `allow_val_change`. On re-entry `tstamp` differs by the ~48 s of startup, so wandb raises `ConfigError`.

The blanket `except Exception` then swallowed it and returned `None`. `WandbBackend.__init__` sets `_run = None` from that (`tracker.py:230`), and `WandbBackend.log()` **early-returns on `_run=None`** (`tracker.py:249`).

Net effect: **a run visible in the W&B UI that silently receives zero metrics for the entire job.** No loss, no MFU, nothing. The early-init change was meant to record *more*; instead it recorded nothing.

Reproduced locally against the shipped code — same `ConfigError`, same `:1706` warning, `call 2 -> None`.

## Fixes

**1. `allow_val_change=True` on the post-init update** (and the `config=` branch). Re-stamping metadata on re-entry is the intended behaviour. Notably, all three later `config.update` sites in `fsdp_tp.py` already passed this flag — `setup_wandb` was the one place that didn't.

**2. Scope the failure handling.** Only the post-init bookkeeping is wrapped in its own `try/except` that keeps the live run. A live run with stale metadata beats a live run with no data.

Deliberately **not** applied to `wandb.init()` itself: with no usable run there's nothing to salvage, and callers rely on `None` to no-op. My first attempt was too broad and broke `test_tracker.py::test_init_failure_all_methods_noop` — that test was right, and a new test now pins the distinction so the handler can't be re-broadened.

## Tests

`tests/test_setup_wandb_reinit.py` — 6 tests, offline, no account needed:

- second call returns a live run (not `None`)
- second call doesn't raise on wall-clock keys
- `log()` works after the second call — the actual end-to-end symptom
- `config=` with a changed value survives re-entry
- post-init failure keeps the run
- **`wandb.init()` failure still returns `None`** (guards the scope)

Verified to bite: all 5 double-init tests fail against shipped v0.24.2.

Full suite: 1383 passed, 1 pre-existing `gqa` failure (the #199 `atol=1e-6` FLASH-vs-MATH issue, identical on clean `main`).

## Summary by Sourcery

Ensure setup_wandb can be safely called multiple times in a single job without breaking WandB logging.

Bug Fixes:
- Allow WandB run config updates to change timestamp and related fields on re-entry without raising errors or dropping the run.
- Scope post-init WandB config update failures so that an already-active run is preserved and metric logging continues, while genuine init failures still return None as before.

Tests:
- Add regression tests verifying setup_wandb survives double initialization, preserves a live run on post-init failures, and still returns None when wandb.init() itself fails.

### CODE REVIEW

File: src/ezpz/distributed.py [hunk 1]
```diff
@@ -1658,48 +1658,84 @@ def setup_wandb(
...
+            # already-created run*. Scoped try/except: a failure here
+            # must not propagate to the outer handler, which returns
+            # None -- that makes WandbBackend set _run=None, and
+            # WandbBackend.log() early-returns on that, so the run stays
+            # visible in the W&B UI while silently receiving zero
+            # metrics for the whole job. A live run with stale metadata
+            # beats a live run with no data. (A failure of wandb.init()
+            # itself is different: there is no usable run, so it still
+            # falls through to the outer handler and returns None.)
+            try:
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/distributed.py:1656]: Confirmed and fixed in 1a43d96 — thanks, this was a real gap.

Verified the premise against `pyproject.toml`: `wandb` is in base `dependencies`, `torch` is an optional extra (`torch = ["torch", "torchaudio", ...]`). So on a torch-less install the post-init `import torch` at what was line 1638 raises with a live run already created, the outer handler returns `None`, and `WandbBackend` discards it — precisely the silent-zero-metrics mode this PR exists to prevent.

The `try` now starts immediately after `wandb.init()` returns a run, covering `run.name`/`run.url`, the bookkeeping imports, the jobid lookup and `config.update()`. Failure of `wandb.init()` itself still falls through to the outer handler and returns `None` (no run to salvage; callers rely on that to no-op).

New test monkeypatches `__import__` to raise on `torch` and asserts the run survives and stays loggable. Verified it actually discriminates: it **fails** against the narrow-scope version and passes against the widened one.

File: tests/test_setup_wandb_reinit.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        )
+        if first is None:
+            pytest.skip("wandb unavailable in this environment")
+        second = setup_wandb(
+            project_name="p", dir=str(tmp_path), mode="offline",
+            config={"lr": 0.003},  # different value, same key
+        )
+        assert second is not None, (
+            "re-supplying config= with a changed value must not kill the run"
+        )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_setup_wandb_reinit.py:119]: Good call — added in 1a43d96. The test now asserts the value actually lands, not just that the run survives.

One correction to the suggested snippet: `setup_wandb` nests the caller's `config=` under `run.config["config"]` (see the `if config is not None: run.config.update({"config": config})` branch), so a top-level `dict(second.config)["lr"]` would be a `KeyError`. The assertion checks the nested location instead:

```python
nested = dict(second.config).get("config")
assert isinstance(nested, dict)
assert nested.get("lr") == 0.003
```

Verified passing — the second call's value does overwrite the first.
